### ISSUE TITLE
Other Unicode characters than Emojis?

### ISSUE THREAD

**Issue by dvergeylen** [rugk/awesome-emoji-picker#109]
Hi rugk,

## Background

I use this add-on on a daily basis and find it frankly awesome. It's super convenient to use and finding appropriate emoji when you are on a desktop computer has never been easier. And as a Thunderbird user, I can't wait for #49 to be implemented!

However, I also use some non emoji unicode characters on a daily basis. Not talking about non ascii characters (Korean, Chinese, ...) but more of symbols like arrows ('⟷', ...), equal signs ('≃', ...), math symbols or simply [checks](https://unicode-table.com/en/sets/check/).

## Proposed solution

It would be convenient to have new tabs listing some additional unicode character sets (there are [plenty of sets](https://unicode-table.com/en/sets/)). Every character has a name just like emojis so searching by name could work as well (it works as an 'answer' in DuckDuckGo when you type "unicode XXX", see [example](https://duckduckgo.com/?q=unicode+left+right+arrow&ia=answer)).

This could be disabled behind a parameter in the options tab as most probably all users won't need such sets (maybe a pick-and-choose manner to select which sets you would like added in your panel? 🤔).

## Alternatives

For now I either make use of DuckDuckGo direct answers when I know the unicode character name by heart (but this can reveal being a difficult exercise when asking for greek letters for instance, they have long names like ['Greek Capital Letter Delta'](https://duckduckgo.com/?q=unicode+Greek+Capital+Letter+Delta&ia=answer) and your are obliged to type the _exact_ unicode name for the answer to show up!).

For some characters, you can have a binding on your keyboard. On belgian keyboards, you can type '≃' by pressing <kbd>Alt Gr</kbd> + <kbd>=</kbd> but that's not the case on french keyboards for instance (although >95% of the layout being the same).

## Additional context

I thought twice before posting this as the Add-on name is "Awesome Emoji pciker", not "Awesome Unicode Characters picker" but watching this repo for a while I have an intuition you aren't against new proposals / discussions.

Greetings from Belgium! 🎉

**Comment by rugk** [comment]
Of course, I'm always for proposals. However, as you also said, the aim of this add-on is to have a picker and other useful features *for emojis*. Other Unicode letters are, as sorry as I am to say this, out-of-scope of this project.
Also they would clutter the user interface, even if it is an optional feature.

I now, how you imagine this, maybe like GNOME Characters does it with separate groups:
![emoji groups](https://user-images.githubusercontent.com/11966684/99390191-80918700-28d8-11eb-93b1-016af9bfe6c3.png)
![character groups](https://user-images.githubusercontent.com/11966684/99390273-a1f27300-28d8-11eb-876d-210af91d96d2.png)

However, I e.g. there don't really like that it mixes colorful emojis with these greyscale arrows or so.
Having a look [at your linked site](https://unicode-table.com/en/sets/) actually most of the groups are emojis. However, the groups atre often not official Unicode groups – even such a group like "Coronavirus". Arrows are, of course, but mixtures like flowers are not. However, you can simply search for "flower" in the emoji picker and you most-likely also get all emojis you want there, the aliases are pretty clever.
As for such advanced research and categorization of emojis (or other chars) really external websites may be best that categorize this to their own preference – which is not standardized. Also [emojipedia e.g. has a corona category](https://emojipedia.org/coronavirus/).

Despite these project-related issues, some technical things: We use [emoji-mart](https://github.com/missive/emoji-mart), which is basically the whole core of the add-on, which itself uses [emoji-data](https://github.com/iamcal/emoji-data) as a data library. As such, we likely would need to have all searchable Unicodes included there, which is unlikely to happen.

**That all said** first of all it's good of you to open this issue! Because I may always reconsider this feature if there is a high-demand (people upvote it or so).
Also, **there is good news**, currently we are in the process of creating a new add-on that deals with other Unicode characters. See https://github.com/rugk/awesome-emoji-picker/issues/106. The new add-on will focus on text conversion at the start, but potentially your idea would be better suited there.

**Comment by dvergeylen** [comment]
Thank you for your prompt response!

I understand this is a "won't fix" as it is out-of-scope of this add-on (was expecting that).
I see you have created a template repo to create WebExtensions for Firefox, this might be a good starting point for me to create a proof-of-concept add-on to mimic this one but with unicode chars? 🤔 Thinking out loud 🙃

**Comment by tdulcet** [comment]
> more of symbols like arrows ('⟷', ...), equal signs ('≃', ...), math symbols or simply [checks](https://unicode-table.com/en/sets/check/).

When #106 is finished (see https://github.com/rugk/unicodify/pull/2), if you install the new add-on, you will be able to type `<-->` and have it autocorrect to ⟷. Same with `~=` to ≅. There are some tables further down [this page](https://tealdulcet.com/autocorrect/) that list all the currently supported autocorrections. Please do let me know if there are any other Unicode autocorrections you think should be included in our new add-on.

> And as a Thunderbird user, I can't wait for #49 to be implemented!

If you do not want to wait, you can install a version of this add-on [here](https://github.com/rugk/awesome-emoji-picker/issues/106#issuecomment-712144781) for testing that does works in Thunderbird and also includes all the features in the new add-on. Just rename the `thunderbirdmanifest.json` file to `manifest.json`. Note that the options page does not load because of [Bug 1641577](https://bugzilla.mozilla.org/show_bug.cgi?id=1641577), which is what #49 is blocked by.

**Comment by dvergeylen** [comment]
Thanks @tdulcet for this information!

> > more of symbols like arrows ('⟷', ...), equal signs ('≃', ...), math symbols or simply [checks](https://unicode-table.com/en/sets/check/).
> 
> When #106 is finished (see [rugk/unicodify#2](https://github.com/rugk/unicodify/pull/2)), if you install the new add-on, you will be able to type `<-->` and have it autocorrect to ⟷. Same with `~=` to ≅. There are some tables further down [this page](https://tealdulcet.com/autocorrect/) that list all the currently supported autocorrections.

Sounds really good, would be glad to be an early adopter! 👍

>Please do let me know if there are any other Unicode autocorrections you think should be included in our new add-on.

There is already a huge list!

> 
> > And as a Thunderbird user, I can't wait for #49 to be implemented!
> 
> If you do not want to wait, you can install a version of this add-on [here](https://github.com/rugk/awesome-emoji-picker/issues/106#issuecomment-712144781) for testing that does work in Thunderbird and it also includes all the features in the new add-on. Just rename the `thunderbirdmanifest.json` file to `manifest.json`. Note that the options page does not load because of [Bug 1641577](https://bugzilla.mozilla.org/show_bug.cgi?id=1641577), which is what #49 is blocked by.

I did download the zip file, moved `thunderbirdmanifest.json` file to `./scripts/manifests/firefox.json` and ran `scripts/make.sh`. It generated a .xpi file successfully but when installing I got an error:

```
Reading manifest: Error processing optional_permissions.1: Value "search" must either: must either [be one of ["clipboardRead", "clipboardWrite", "geolocation", "idle", "notifications"], be one of ["accountsRead"], be one of ["addressBooks"], be one of ["accountsFolders"], be one of ["menus.overrideContext"], be one of ["messagesMove", "messagesRead"], be one of ["activeTab", "tabs", "tabHide"], be one of ["browserSettings"], be one of ["cookies"], be one of ["downloads", "downloads.open"], be one of ["topSites"], be one of ["webNavigation"], or be one of ["webRequest", "webRequestBlocking"]], or must either [be one of ["<all_urls>"], must either [match the pattern /^(https?|wss?|file|ftp|\*):\/\/(\*|\*\.[^*/]+|[^*/]+)\/.*$/, or match the pattern /^file:\/\/\/.*$/], or match the pattern /^resource:\/\/(\*|\*\.[^*/]+|[^*/]+)\/.*$|^about:/]
Reading manifest: Error processing permissions.3: Value "contextMenus" must either: must either [must either [be one of ["clipboardRead", "clipboardWrite", "geolocation", "idle", "notifications"], be one of ["accountsRead"], be one of ["addressBooks"], be one of ["accountsFolders"], be one of ["menus.overrideContext"], be one of ["messagesMove", "messagesRead"], be one of ["activeTab", "tabs", "tabHide"], be one of ["browserSettings"], be one of ["cookies"], be one of ["downloads", "downloads.open"], be one of ["topSites"], be one of ["webNavigation"], or be one of ["webRequest", "webRequestBlocking"]], be one of ["alarms", "mozillaAddons", "storage", "unlimitedStorage"], be one of ["geckoProfiler"], be one of ["menus"], be one of ["pkcs11"], be one of ["contextualIdentities"], be one of ["dns"], be one of ["management"], be one of ["privacy"], be one of ["proxy"], be one of ["nativeMessaging"], be one of ["telemetry"], be one of ["theme"], or match the pattern /^experiments(\.\w+)+$/], or must either [be one of ["<all_urls>"], must either [match the pattern /^(https?|wss?|file|ftp|\*):\/\/(\*|\*\.[^*/]+|[^*/]+)\/.*$/, or match the pattern /^file:\/\/\/.*$/], or match the pattern /^resource:\/\/(\*|\*\.[^*/]+|[^*/]+)\/.*$|^about:/]
Reading manifest: Error processing omnibox: An unexpected property was found in the WebExtension manifest.
```

It did not find the "search" permission (Thunderbird 68.12). Maybe I should open a new issue elsewhere?

**Comment by tdulcet** [comment]
> Thanks @tdulcet for this information!

No problem!

> It did not find the "search" permission (Thunderbird 68.12)

The "search" permission is only listed in the Firefox manifest, so I believe you are somehow using the wrong one. Also, I should have mentioned that it requires Thunderbird 78, the [current ESR version](https://www.thunderbird.net/).

**Comment by rugk** [comment]
Thanks all for chiming in here. That said, I'd kindly ask you to please move your discussion regarding Thunderbird (tests) into the issue #49 about Thunderbird, actually. :smiley:
